gluten-core/src/main/scala/io/glutenproject/execution/WriteFilesExecTransformer.scala

```diff
@@ -0,0 +1,173 @@
+import org.apache.spark.sql.catalyst.expressions.Attribute
+import org.apache.spark.sql.execution.{SparkPlan, UnaryExecNode}
+import org.apache.spark.sql.execution.datasources.FileFormat
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+import com.google.protobuf.{Any, StringValue}
+
+import scala.collection.JavaConverters._
+import scala.collection.convert.ImplicitConversions.`collection AsScalaIterable`
+
+case class WriteFilesExecTransformer(
+    child: SparkPlan,
+    fileFormat: FileFormat,
+    partitionColumns: Seq[Attribute],
+    bucketSpec: Option[BucketSpec],
+    options: Map[String, String],
+    staticPartitions: TablePartitionSpec)
+  extends UnaryExecNode
+  with UnaryTransformSupport {
+  override def metricsUpdater(): MetricsUpdater = null
+
+  override def output: Seq[Attribute] = Seq.empty
+
+  def genWriteParameters(): Any = {
+    val compressionCodec = if (options.get("parquet.compression").nonEmpty) {
+      options.get("parquet.compression").get.toLowerCase().capitalize
+    } else SQLConf.get.parquetCompressionCodec.toLowerCase().capitalize
+
+    val writeParametersStr = new StringBuffer("WriteParameters:")
+    writeParametersStr.append("is").append(compressionCodec).append("=1").append("\n")
+    val message = StringValue
+      .newBuilder()
+      .setValue(writeParametersStr.toString)
+      .build()
+    BackendsApiManager.getTransformerApiInstance.getPackMessage(message)
+  }
+
+  def createEnhancement(output: Seq[Attribute]): com.google.protobuf.Any = {
+    val inputTypeNodes = output.map {
+      attr => ConverterUtils.getTypeNode(attr.dataType, attr.nullable)
+    }
+
+    BackendsApiManager.getTransformerApiInstance.getPackMessage(
+      TypeBuilder.makeStruct(false, inputTypeNodes.asJava).toProtobuf)
+  }
+
+  def getRelNode(
+      context: SubstraitContext,
+      originalInputAttributes: Seq[Attribute],
+      writePath: String,
+      operatorId: Long,
+      input: RelNode,
+      validation: Boolean): RelNode = {
+    val typeNodes = ConverterUtils.collectAttributeTypeNodes(originalInputAttributes)
+
+    val columnTypeNodes = new java.util.ArrayList[ColumnTypeNode]()
+    val inputAttributes = new java.util.ArrayList[Attribute]()
+    val childSize = this.child.output.size
+    val childOutput = this.child.output
+    for (i <- 0 until childSize) {
+      val partitionCol = partitionColumns.find(_.exprId == childOutput(i).exprId)
+      if (partitionCol.nonEmpty) {
+        columnTypeNodes.add(new ColumnTypeNode(1))
+        // "aggregate with partition group by can be pushed down"
+        // test partitionKey("p") is different with
+        // data columns("P").
+        inputAttributes.add(partitionCol.get)
+      } else {
+        columnTypeNodes.add(new ColumnTypeNode(0))
+        inputAttributes.add(originalInputAttributes(i))
+      }
+    }
+
+    val nameList =
+      ConverterUtils.collectAttributeNames(inputAttributes.toSeq)
```

Comment:
It's risky to use the child output name to write parquet file. I think we should use the output columns of WriteFiles to match the table column.

Reply:
@ulysses-you It seems the output of WriteFiles is empty. https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/WriteFiles.scala#L71

Reply:
the output columans is at `WriteFilesSpec.WriteJobDescription.allColumns`